pkg/trace/api/proxy.go

```diff
@@ -0,0 +1,192 @@
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/trace/logutil"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type proxyEndpointsConfig struct {
+	name                      string
+	enabledConfig             string // config entry for enabling disabling proxy endpoint, can be empty
+	ddURLConfig               string // config entry for intake url *_DD_URL
+	additionalEndpointsConfig string // config entry for adding additional, can be empty
+	urlTemplate               string // template to build intake URL based on SITE config
+	defaultURL                string
+	additionalDirector        func(req *http.Request)
+}
+
+func (pec *proxyEndpointsConfig) proxyRoundTripper(client http.RoundTripper, apiKey string) (http.RoundTripper, error) {
+	urls, apiKeys, err := pec.proxyEndpoints(apiKey)
+	return &proxyMultiTransport{
+		rt:      client,
+		targets: urls,
+		keys:    apiKeys,
+		baseTag: "type:" + pec.name,
+	}, err
+
+}
+func (pec *proxyEndpointsConfig) proxyEndpoints(apiKey string) (urls []*url.URL, apiKeys []string, err error) {
+	main := pec.defaultURL
+	if v := config.Datadog.GetString(pec.ddURLConfig); v != "" {
+		main = v
+	} else if site := config.Datadog.GetString("site"); site != "" {
+		main = fmt.Sprintf(pec.urlTemplate, site)
+	}
+	u, err := url.Parse(main)
+	if err != nil {
+		// if the main intake URL is invalid we don't use additional endpoints
+		return nil, nil, fmt.Errorf("error parsing main %s intake URL %s: %v", pec.name, main, err)
+	}
+	urls = append(urls, u)
+	apiKeys = append(apiKeys, apiKey)
+
+	if pec.additionalEndpointsConfig != "" && config.Datadog.IsSet(pec.additionalEndpointsConfig) {
+		extra := config.Datadog.GetStringMapStringSlice(pec.additionalEndpointsConfig)
+		for endpoint, keys := range extra {
+			u, err := url.Parse(endpoint)
+			if err != nil {
+				log.Errorf("Error parsing additional %s intake URL %s: %v", pec.name, endpoint, err)
+				continue
+			}
+			for _, key := range keys {
+				urls = append(urls, u)
+				apiKeys = append(apiKeys, key)
+			}
+		}
+	}
+	return urls, apiKeys, nil
+}
+
+// forwardingProxyHandler returns a new HTTP handler which will proxy requests to the configured intakes.
+// If the main intake URL can not be computed because of config, the returned handler will always
+// return http.StatusInternalServerError along with a clarification.
+func (r *HTTPReceiver) forwardingProxyHandler(pec *proxyEndpointsConfig) http.Handler {
+	if pec.enabledConfig != "" && !config.Datadog.GetBool(pec.enabledConfig) {
+		return pec.proxyDisabledHandler(nil)
+	}
+	transport, err := pec.proxyRoundTripper(r.conf.NewHTTPTransport(), r.conf.APIKey())
+	if err != nil {
+		return pec.proxyDisabledHandler(err)
+	}
+
+	return pec.newForwardingProxy(transport, r.conf.Hostname, r.conf.DefaultEnv)
```

Comment:
This should be a function, not a method. The proxy config `pec` should be an argument not a receiver. The return values with accessing private functions like `pec.proxyDisabledHandler` and passing them errors from an outer scope is an odd pattern.

I think this could be more idiomatic and simplified if we got rid of this entire function (`(*HTTPReceiver).forwardingProxyHandler`) and instead just return one call in `endpoints.go`:
```go
return newForwardingProxy(pec, r.conf.Hostname, r.conf.DefaultEnv)
```
This function will then return an `http.Handler` which could be the proxy or the "disabled handler".